File: ontology-driven-dev/code-app-example/backend/api/ai.py
import json

from flask import Blueprint, Response, g, request

from ai import chat, config as ai_config
from services import auth_service
from utils.response import fail, ok
from utils.security import login_required

bp = Blueprint("ai", __name__, url_prefix="/api/ai")


def _is_admin():
    return "*" in auth_service.get_permission_codes(getattr(g, "user_id", None))


@bp.get("/config")
@login_required
def get_config():
    cfg = ai_config.get_config()
    if not _is_admin():
        return fail("无权限"), 403
    # 掩码 api_key，避免泄露
    api_key = cfg["api_key"]
    masked = ""
    if api_key:
        masked = api_key[:4] + "****" + api_key[-4:] if len(api_key) > 8 else "****"
    return ok({
        "base_url": cfg["base_url"],
        "api_key_masked": masked,
        "model_id": cfg["model_id"],
        "max_tokens": cfg["max_tokens"],
        "configured": cfg["configured"],
    })


@bp.post("/config")
@login_required
def save_config():
    if not _is_admin():
        return fail("无权限"), 403
    data = request.get_json(silent=True) or {}
    cfg = ai_config.save_config(data, auth_service.current_user())
    return ok({"configured": cfg["configured"]}, "保存成功")


@bp.post("/test")
@login_required
def test_connection():
    if not _is_admin():
        return fail("无权限"), 403
    data = request.get_json(silent=True) or {}
    result = ai_config.test_config(data)
    return ok(result)


@bp.post("/chat")
@login_required
def ai_chat():
    data = request.get_json(silent=True) or {}
    message = (data.get("message") or "").strip()
    history = data.get("history") or []
    if not message:
        return fail("请输入您的问题")
    user = auth_service.current_user()

    def generate():
        try:
            for event, payload in chat.run_chat(message, history, user):
                yield f"event: {event}\ndata: {json.dumps(payload, ensure_ascii=False)}\n\n"
        except Exception as e:
            yield f"event: delta\ndata: {json.dumps({'content': f'系统错误：{e}'}, ensure_ascii=False)}\n\n"
            yield "event: message_end\ndata: {}\n\n"
        finally:
            auth_service.write_audit(user["id"], user["username"], "AI_CHAT", message[:200])

    return Response(
        generate(),
        mimetype="text/event-stream",
        headers={"Cache-Control": "no-cache", "X-Accel-Buffering": "no"},
    )

File: ontology-driven-dev/code-app-example/frontend/src/api/ai.ts
import { getToken, http } from './request'

export interface RenderPayload {
  type: 'text' | 'table' | 'chart' | 'action'
  content?: string
  columns?: string[]
  rows?: Record<string, any>[]
  label?: string
  path?: string
  chart_type?: string
  title?: string
  data?: Record<string, any>[]
}

export interface AiConfig {
  base_url: string
  api_key_masked: string
  model_id: string
  max_tokens: number
  configured: boolean
}

export interface TestResult {
  success: boolean
  latency_ms: number
  reply: string
  message: string
}

export interface ChatCallbacks {
  onDelta: (text: string) => void
  onRender: (payload: RenderPayload) => void
  onToolCall: (name: string) => void
  onDone: () => void
  onError: (msg: string) => void
}

export const aiApi = {
  getConfig() {
    return http.get<AiConfig>('/ai/config')
  },
  saveConfig(data: { base_url: string; api_key: string; model_id: string; max_tokens: number }) {
    return http.post('/ai/config', data)
  },
  testConnection(data: { base_url: string; api_key: string; model_id: string; max_tokens: number }) {
    return http.post<TestResult>('/ai/test', data)
  },
  async chat(message: string, history: { role: string; content: string }[], cb: ChatCallbacks) {
    const token = getToken()
    let resp: Response
    try {
      resp = await fetch('/api/ai/chat', {
        method: 'POST',
        headers: { 'Content-Type': 'application/json', Authorization: `Bearer ${token}` },
        body: JSON.stringify({ message, history }),
      })
    } catch (e: any) {
      cb.onError('网络错误：' + e.message)
      cb.onDone()
      return
    }
    if (!resp.ok || !resp.body) {
      const body = await resp.json().catch(() => null)
      cb.onError(body?.message || `请求失败（HTTP ${resp.status}）`)
      cb.onDone()
      return
    }
    const reader = resp.body.getReader()
    const decoder = new TextDecoder()
    let buffer = ''
    let eventType = ''
    while (true) {
      const { done, value } = await reader.read()
      if (done) break
      buffer += decoder.decode(value, { stream: true })
      const lines = buffer.split('\n')
      buffer = lines.pop() || ''
      for (const line of lines) {
        const l = line.trim()
        if (l.startsWith('event:')) {
          eventType = l.slice(6).trim()
        } else if (l.startsWith('data:')) {
          const data = l.slice(5).trim()
          if (!data) continue
          let payload: any = {}
          try { payload = JSON.parse(data) } catch { continue }
          if (eventType === 'delta') cb.onDelta(payload.content || '')
          else if (eventType === 'render_payload') cb.onRender(payload)
          else if (eventType === 'tool_call') cb.onToolCall(payload.name || '')
          else if (eventType === 'message_end') cb.onDone()
        }
      }
    }
    cb.onDone()
  },
}
